src/ledger/LedgerStateSnapshot.h

```diff
@@ -0,0 +1,179 @@
+    LedgerEntryWrapper getSourceAccount(LedgerHeaderWrapper const& header,
+                                        TransactionFrame const& tx) override;
+    LedgerEntryWrapper getSourceAccount(LedgerHeaderWrapper const& header,
+                                        TransactionFrame const& tx,
+                                        AccountID const& AccountID) override;
+    LedgerEntryWrapper load(LedgerKey const& key) override;
+#ifdef BUILD_TESTS
+    LedgerEntryWrapper loadToModify(LedgerKey const& key) override;
+#endif
+    AbstractLedgerTxn&
+    getLedgerTxn() const
+    {
+        return mLedgerTxn;
+    }
+};
+
+// A concrete implementation of read-only BucketList snapshot wrapper
+class BucketSnapshotState : public AbstractLedgerStateSnapshot
+{
+    BucketManager& mBucketManager;
+    std::shared_ptr<SearchableBucketListSnapshot> mSnapshot;
+
+  public:
+    BucketSnapshotState(BucketManager& bm);
+    ~BucketSnapshotState() override;
+
+    LedgerHeaderWrapper getLedgerHeader() override;
+    LedgerHeaderWrapper getHeader() override;
+    LedgerEntryWrapper getSourceAccount(AccountID const& account) override;
+    LedgerEntryWrapper getSourceAccount(LedgerHeaderWrapper const& header,
+                                        TransactionFrame const& tx) override;
+    LedgerEntryWrapper getSourceAccount(LedgerHeaderWrapper const& header,
+                                        TransactionFrame const& tx,
+                                        AccountID const& AccountID) override;
+    LedgerEntryWrapper load(LedgerKey const& key) override;
+#ifdef BUILD_TESTS
+    LedgerEntryWrapper
+    loadToModify(LedgerKey const& key) override
+    {
+        return load(key);
+    }
+#endif
+};
+
+// A helper class to create and query read-only snapshots
+// Automatically decides whether to create a BucketList (recommended), or SQL
+// snapshot (deprecated, but currently supported)
+class LedgerStateGetter : public NonMovableOrCopyable
+{
+    std::shared_ptr<AbstractLedgerStateSnapshot> mGetter;
+    std::unique_ptr<LedgerTxn> mLegacyLedgerTxn;
+
+  public:
+    LedgerStateGetter(AbstractLedgerTxn& ltx);
+    LedgerStateGetter(Application& app);
+    LedgerHeaderWrapper getHeader();
+    LedgerHeaderWrapper getLedgerHeader();
+    LedgerEntryWrapper getSourceAccount(AccountID const& account);
+    LedgerEntryWrapper
+    getSourceAccount(LedgerHeaderWrapper const& header,
+                     TransactionFrame const& tx)
+    {
+        return mGetter->getSourceAccount(header, tx);
+    }
+    LedgerEntryWrapper
+    getSourceAccount(LedgerHeaderWrapper const& header,
+                     TransactionFrame const& tx, AccountID const& AccountID)
+    {
+        return mGetter->getSourceAccount(header, tx, AccountID);
+    }
+    LedgerEntryWrapper load(LedgerKey const& key);
+#ifdef BUILD_TESTS
+    LedgerEntryWrapper loadToModify(LedgerKey const& key);
+#endif
+    std::shared_ptr<AbstractLedgerStateSnapshot>
+    getSnapshot() const
```

Comment:
good idea. LedgerSnapshot is not copyable or movable, so I added `executeWithMaybeInnerSnapshot` to the interface which takes a lambda and creates a nested LedgerTxn in `LedgerTxnReadOnly`. It also throws in `BucketSnapshotState`, since we shouldn't be nesting bucketlist snapshots.